edb/schema/referencing.py

```diff
@@ -38,12 +38,26 @@
 ReferencedT = TypeVar('ReferencedT', bound='ReferencedObject')
 
 
+if TYPE_CHECKING:
+    from edb.schema import inheriting as s_inheriting
+
+
 class ReferencedObject(so.Object, derivable.DerivableObjectBase):
 
-    def get_referrer(self, schema):
+    def get_referrer(self, schema: s_schema.Schema) -> Optional[so.Object]:
```

Comment:
This is my biggest doubt of my current PR. 

In practice, the `ReferencedObject` class is inherited by classes that all define a `subject` field: therefore they all have a `get_subject` method independently from each other (s_annos.AnnotationValue, s_indexes.Index, s_constraints.Constraint).

In this kind of situation, I would define the `get_subject` as an `@abstractmethod` on the `ReferencedObject` class, to make it explicit that implementers declare such method - and make both humans reading the code and mypy happy.

Since the `get_subject` method is generated dynamically from the `subject = SchemaField`, I think abstractmethod doesn't fit here. I also noticed that you don't use abstractmethod, so I think you don't like that pattern.

I also considered specifying a class property:

```
get_subject: Callable[[ReferencedObject, s_schema.Schema], Optional[so.Object]]
```

But this looks confusing, too, and misleading for humans who read the code. Asserting that self is an instance of the classes having a subject, for the method that requires `get_subject`, violates the Single Responsibility Principle but looks like the clearest solution.

Reply:
A regular method raising a `NotImplementedError` should suffice.  I agree that the implementation should be moved to the classes that actually declare `subject`.

Reply:
Super :blush: in this case, I try to add a method raising `NotImplementedError` to the class, and I see if all tests pass. I am correcting another point right now.